include/fi_mem.h

```diff
@@ -81,6 +81,22 @@ do {								\
 } while (0)
 #define freestack_pop(fs) freestack_pop_impl(fs, (fs)->next)
 
+#define freestack_shm_push(fs, p_offset, offset) freestack_push(fs, \
+	freestack_to_address(fs, p_offset, offset))		\
+
+#define freestack_shm_pop(fs, offset) freestack_to_offset(fs,	\
+	freestack_pop_impl(fs, &(fs)->next), offset)		\
+
+static inline void *freestack_to_address(void *fs, size_t p_off, size_t fs_off)
+{
+	return (char **) fs + (p_off - fs_off);
+}
+
+static inline size_t freestack_to_offset(void *fs, void *p, size_t fs_off)
+{
+	return ((char **) p - (char **) fs) + fs_off;
+}
+
```

Comment:
I'd like to see the shm freestack api more closely mirror the normal freestack api.  For example, freestack_shm_pop() should return a valid pointer to the popped item.  freestack_shm_push() should take that pointer as input.  Pushing/popping offsets is odd.

What's needed is a generic way to map an address (pointer) from one process to another.  I don't think there's a way to do this.  Adding 'struct smr_region *base_addr' to smr_region would allow this.  The process that creates the smr_region can set base_addr to the start of the region.  Any peer can compare their smr_region pointer to the base_addr to determine the offset needed to convert pointers between the processes.